File: src/profiling/summary.ts
import type { ProfileReport } from "./types.js";

/**
 * Format a profile report as a human-readable summary for console output.
 */
export function formatProfileSummary(report: ProfileReport): string {
  const lines: string[] = [];
  const total = report.meta.totalDurationMs;

  lines.push("=== Performance Profile ===");
  lines.push(`Total duration: ${formatDuration(total)}`);
  lines.push("");

  // Stage breakdown
  if (report.stageSummaries.length > 0) {
    lines.push("Stage breakdown:");
    // Sort by duration descending
    const sorted = [...report.stageSummaries].sort((a, b) => b.durationMs - a.durationMs);
    for (const stage of sorted) {
      const pct = total > 0 ? ((stage.durationMs / total) * 100).toFixed(1) : "0.0";
      lines.push(`  ${stage.name.padEnd(25)} ${formatDuration(stage.durationMs).padStart(10)}  (${pct}%)${stage.spanCount > 1 ? `  [${stage.spanCount} spans]` : ""}`);
    }
    lines.push("");
  }

  // Rename timing percentiles
  if (report.renameTiming) {
    const rt = report.renameTiming;
    lines.push(`Rename timing (${rt.count} functions):`);
    lines.push(`  p50: ${formatDuration(rt.p50)}  p95: ${formatDuration(rt.p95)}  p99: ${formatDuration(rt.p99)}`);
    lines.push(`  min: ${formatDuration(rt.minMs)}  max: ${formatDuration(rt.maxMs)}`);
    lines.push("");
  }

  // Concurrency stats
  if (report.concurrencySnapshots.length > 0) {
    const snapshots = report.concurrencySnapshots;
    const avgInFlight = snapshots.reduce((s, c) => s + c.inFlight, 0) / snapshots.length;
    const maxInFlight = Math.max(...snapshots.map(c => c.inFlight));
    const avgReady = snapshots.reduce((s, c) => s + c.ready, 0) / snapshots.length;
    lines.push("Concurrency utilization:");
    lines.push(`  avg in-flight: ${avgInFlight.toFixed(1)}  max in-flight: ${maxInFlight}`);
    lines.push(`  avg ready: ${avgReady.toFixed(1)}  samples: ${snapshots.length}`);
    lines.push("");
  }

  return lines.join("\n");
}

function formatDuration(ms: number): string {
  if (ms < 1000) return `${ms.toFixed(0)}ms`;
  if (ms < 60_000) return `${(ms / 1000).toFixed(1)}s`;
  const mins = Math.floor(ms / 60_000);
  const secs = ((ms % 60_000) / 1000).toFixed(0);
  return `${mins}m${secs}s`;
}

File: crates/humanify-core/src/profiling/summary.rs
//! The console profile summary (TS: src/profiling/summary.ts). Its JS
//! number text — `formatDuration`, `Number.prototype.toString`, `toFixed`
//! — comes from the one owner, `humanify_model::js`. Byte-equal to the TS
//! on the frozen vectors (test/parity/wpb5-profile-vectors.json).

use std::fmt::Write as _;

use humanify_model::js::{format_duration, to_fixed as js_to_fixed, utf16_len};
use humanify_model::profiling::{ProfileReport, StageSummary};

/// `s.padEnd(n)` / `s.padStart(n)` — widths in UTF-16 code units.
fn pad_end(s: &str, n: usize) -> String {
    format!("{s}{}", " ".repeat(n.saturating_sub(utf16_len(s))))
}

fn pad_start(s: &str, n: usize) -> String {
    format!("{}{s}", " ".repeat(n.saturating_sub(utf16_len(s))))
}

fn stage_lines(stages: &[StageSummary], total: f64, lines: &mut Vec<String>) {
    lines.push("Stage breakdown:".to_string());
    let mut sorted: Vec<&StageSummary> = stages.iter().collect();
    // Stable, descending — ties keep report order (JS Array.sort is stable).
    sorted.sort_by(|a, b| {
        b.duration_ms
            .0
            .partial_cmp(&a.duration_ms.0)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    for stage in sorted {
        let pct = if total > 0.0 {
            js_to_fixed(stage.duration_ms.0 / total * 100.0, 1)
        } else {
            "0.0".to_string()
        };
        let mut line = format!(
            "  {} {}  ({pct}%)",
            pad_end(&stage.name, 25),
            pad_start(&format_duration(stage.duration_ms.0), 10)
        );
        if stage.span_count > 1 {
            let _ = write!(line, "  [{} spans]", stage.span_count);
        }
        lines.push(line);
    }
    lines.push(String::new());
}

/// TS `formatProfileSummary`.
pub fn format_profile_summary(report: &ProfileReport) -> String {
    let total = report.meta.total_duration_ms.0;
    let mut lines = vec![
        "=== Performance Profile ===".to_string(),
        format!("Total duration: {}", format_duration(total)),
        String::new(),
    ];

    if !report.stage_summaries.is_empty() {
        stage_lines(&report.stage_summaries, total, &mut lines);
    }

    if let Some(rt) = &report.rename_timing {
        lines.push(format!("Rename timing ({} functions):", rt.count));
        lines.push(format!(
            "  p50: {}  p95: {}  p99: {}",
            format_duration(rt.p50.0),
            format_duration(rt.p95.0),
            format_duration(rt.p99.0)
        ));
        lines.push(format!(
            "  min: {}  max: {}",
            format_duration(rt.min_ms.0),
            format_duration(rt.max_ms.0)
        ));
        lines.push(String::new());
    }

    let snaps = &report.concurrency_snapshots;
    if !snaps.is_empty() {
        #[allow(clippy::cast_precision_loss)]
        let n = snaps.len() as f64;
        let avg_in_flight = snaps.iter().fold(0.0, |s, c| s + f64::from(c.in_flight)) / n;
        let max_in_flight = snaps.iter().map(|c| c.in_flight).max().unwrap_or(0);
        let avg_ready = snaps.iter().fold(0.0, |s, c| s + f64::from(c.ready)) / n;
        lines.push("Concurrency utilization:".to_string());
        lines.push(format!(
            "  avg in-flight: {}  max in-flight: {max_in_flight}",
            js_to_fixed(avg_in_flight, 1)
        ));
        lines.push(format!(
            "  avg ready: {}  samples: {}",
            js_to_fixed(avg_ready, 1),
            snaps.len()
        ));
        lines.push(String::new());
    }

    lines.join("\n")
}
